Source/WebKit/WebProcess/Cache/WebCacheStorageConnection.cpp

```diff
@@ -38,19 +38,33 @@
 
 namespace WebKit {
 
-WebCacheStorageConnection::WebCacheStorageConnection(WebCacheStorageProvider& provider)
-    : m_provider(provider)
+WebCacheStorageConnection::WebCacheStorageConnection()
 {
 }
 
 WebCacheStorageConnection::~WebCacheStorageConnection()
 {
-    ASSERT(isMainRunLoop());
 }
 
-IPC::Connection& WebCacheStorageConnection::connection()
+Ref<IPC::Connection> WebCacheStorageConnection::connection()
 {
-    return WebProcess::singleton().ensureNetworkProcessConnection().connection();
+    {
+        Locker lock(m_connectionLock);
+        if (m_connection)
+            return *m_connection;
+    }
+
+    RefPtr<IPC::Connection> connection;
+    callOnMainRunLoopAndWait([&connection]() mutable {
+        connection = &WebProcess::singleton().ensureNetworkProcessConnection().connection();
+    });
+
+    {
+        Locker lock(m_connectionLock);
+        m_connection = connection;
+    }
+
+    return connection.releaseNonNull();
```

Comment:
Is it safe to not hold lock between check and set? What will happen in the following case:
```
T1: hold lock -> check m_connection -> release lock
Main: create connection
Main: connection is closed, hold lock -> set m_connection null -> release lock
T1: hold lock -> set m_connection with connection -> release lock
```

Reply:
Right, let's only set m_connection in main thread to prevent this issue.